File: DSA/src/Questions/Array/MinimumAreaRect.java
package Questions.Array;

import java.util.HashSet;
import java.util.Set;

/*
https://leetcode.com/problems/minimum-area-rectangle/description
You are given an array of points in the X-Y plane points where points[i] = [xi, yi].

Return the minimum area of a rectangle formed from these points, with sides parallel to the X and Y axes.
If there is not any such rectangle, return 0.


Example 1:


Input: points = [[1,1],[1,3],[3,1],[3,3],[2,2]]
Output: 4
Example 2:


Input: points = [[1,1],[1,3],[3,1],[3,3],[4,1],[4,3]]
Output: 2

Constraints:

1 <= points.length <= 500
points[i].length == 2
0 <= xi, yi <= 4 * 104
All the given points are unique.
 */
public class MinimumAreaRect {
    public static double minAreaFreeRect(int[][] points) {
        int n = points.length;
        double area = Double.MAX_VALUE;
        Set<String> pointSet = new HashSet<>();
        for (int[] point : points) {
            pointSet.add(point[0] + "," + point[1]);
        }
        for (int i = 0; i < n; i++) {
            for (int j = i + 1; j < n; j++) {
                for (int k = j + 1; k < n; k++) {
                    // Check if the points are diagonals of a rectangle
                    int dx1 = points[k][0] - points[i][0];
                    int dy1 = points[k][1] - points[i][1];
                    int dx2 = points[j][0] - points[i][0];
                    int dy2 = points[j][1] - points[i][1];
                    // Dot product of the edges are 0 when perpendicular
                    /*
                     * If the dot product of the two edges is 0, then the two edges are
                     * perpendicular to each other.
                     */
                    if (dy1 * dy2 + dx1 * dx2 == 0) {
                        int xl = points[j][0] + dx1;
                        int yl = points[j][1] + dy1;
                        if (pointSet.contains(xl + "," + yl)) {
                            area = Math.min(area, Math.sqrt(dx1 * dx1 + dy1 * dy1) * Math.sqrt(dx2 * dx2 + dy2 * dy2));
                        }
                    }
                }
            }
        }
        return area == Double.MAX_VALUE ? 0 : area;
    }


    public static void main(String[] args) {
        int[][] points = { { 1, 1 }, { 1, 3 }, { 3, 1 }, { 3, 3 }, { 2, 2 } };
        System.out.println(minAreaFreeRect(points));
    }
}

File: DSA/Python/Questions/Array/MinimumAreaRect.py
"""
# Question: Minimum Area Rectangle
# Link: https://leetcode.com/problems/minimum-area-rectangle/

# Given points in X-Y plane, find minimum area of rectangle with sides parallel to axes.
# Return 0 if no such rectangle exists.

# Time Complexity: O(n^3) where n is number of points
# Space Complexity: O(n) for storing point set

# Algorithm:
# 1. Convert points to set for O(1) lookup
# 2. Check all possible triplets of points
# 3. Verify if fourth point exists to form rectangle
# 4. Calculate and track minimum area

# Key Components:
# - min_area_free_rect(): Main function implementing area calculation
# - Point set for efficient lookup
# - Perpendicular check using dot product
"""

from typing import List
import math


class MinimumAreaRect:
    def min_area_free_rect(self, points: List[List[int]]) -> float:
        n = len(points)
        area = float("inf")
        point_set = {f"{x},{y}" for x, y in points}

        for i in range(n):
            for j in range(i + 1, n):
                for k in range(j + 1, n):
                    # Check if points form perpendicular edges
                    dx1 = points[k][0] - points[i][0]
                    dy1 = points[k][1] - points[i][1]
                    dx2 = points[j][0] - points[i][0]
                    dy2 = points[j][1] - points[i][1]

                    # Dot product check for perpendicular edges
                    if dy1 * dy2 + dx1 * dx2 == 0:
                        xl = points[j][0] + dx1
                        yl = points[j][1] + dy1

                        if f"{xl},{yl}" in point_set:
                            edge1 = math.sqrt(dx1 * dx1 + dy1 * dy1)
                            edge2 = math.sqrt(dx2 * dx2 + dy2 * dy2)
                            current_area = edge1 * edge2
                            area = min(area, current_area)

        return 0 if area == float("inf") else area


def main():
    solution = MinimumAreaRect()
    points = [[1, 1], [1, 3], [3, 1], [3, 3], [2, 2]]
    result = solution.min_area_free_rect(points)
    print(result)  # Expected output: 4.0


if __name__ == "__main__":
    main()
